### ISSUE TITLE
Copied Text that includes a File URI ( ``file:// or file:/// ``) is Not Pasted

### ISSUE THREAD

**Issue by ZtzEAcRsND** [zim-desktop-wiki/zim-desktop-wiki#1598]
#### Summary
Attempt to copy or cut text with a file URI (``file:// or file:///) from`` one zim page, fails to paste to different location on the same zim page or to a different zim page

#### Reproduce

* Create new zim note book
* Create page1 and page2
* Create a file URI by dragging a directory (e.g., thunar) icon onto the zim page1 (e.g., ``/etc``) from which zim creates a URI '``file:///etc``' 
* Copy the file URI on page1 (either as plain 'copy' or 'copy link' or any 'copy as' option. Attempt to paste the file URI to the bottom of page 1
	* expected result: file URI is pasted
	* actual result: no pasted file URI (either as plain 'paste' or as 'verbatim')
* Copy the file URI on page1, attempt to paste the file URI to zim page2
	* expected result: file URI is pasted
	* actual result: no pasted file URI (either as plain 'paste' or as 'verbatim')
* Similar issues occur if the problematic file URI is inside a block of variously formatted text.  Result of attempt cut/copy/paste: no text with the embedded file URI is pasted. 

#### Background / Related to

* tests run zim 0.74.0, zim started directly with zim.py from zim package extracted to user directory [``~/.local/lib,``](file:///home/walter/.local/lib%2C) on xubuntu 21.04
* appears to NOT be a problem in zim 0.73.5 (xubuntu 21.04) or 0.73.4 (win10)
* may be similar to issue #1412
* may be similar to issue #862
* in #862 JK points to additional information: <https://datatracker.ietf.org/doc/html/rfc8089>

#### More Detail with Various Forms of the URI file:<something>etc

For simplicity, all constructs of "file:<something>" may be referred to as "file URI" even though 'wiki syntax' might be more correct. The case results may be the same without the 'etc' following 'file:'  The two main problem cases for cut/copy/paste are "``file://``" and "``file:///``". Other issues (not specifically cut/copy/paste) are noted for those two and other cases (noted as "[possible issue]". 

case ==>  ``file:etc`` 

* pastes as "``file:etc"`` with <ctrl>c or context menu [<right click>paste ] into another zim page with auto-link added when copied as link or not link or as verbatim format
* copy/cut, then paste works as expected
* mouse click on file URI opens empty zim page ``file:etc``
* pop up shows "Go to file:etc"
* [``/tmp/zim-<random``](file:///tmp/zim-%3Crandom)``>/zim.log relevant text after copy/paste:``

``DEBUG: Targets available for paste: [Gdk.Atom.intern("TIMESTAMP", False), Gdk.Atom.intern("TARGETS", False), Gdk.Atom.intern("MULTIPLE", False), Gdk.Atom.intern("SAVE_TARGETS", False), Gdk.Atom.intern("UTF8_STRING", False), Gdk.Atom.intern("COMPOUND_TEXT", False), Gdk.Atom.intern("TEXT", False), Gdk.Atom.intern("STRING", False), Gdk.Atom.intern("text/plain;charset=utf-8", False), Gdk.Atom.intern("text/plain", False)], we want parsetree``

case ==>  ``file:/etc ``

* pastes as "``file:etc"`` with <ctrl>c or context menu [<right click>paste ] into another zim page with auto-link added when copied as link or not link or as verbatim format
* pop up shows "``open file:///etc"``
* mouse click on file URI opens empty zim page ``file:etc [different than pop up] [possible issue]``
* [``/tmp/zim-<random``](file:///tmp/zim-%3Crandom)``>/zim.log relevant text after copy/paste:``

``DEBUG: Targets available for paste: [Gdk.Atom.intern("TIMESTAMP", False), Gdk.Atom.intern("TARGETS", False), Gdk.Atom.intern("MULTIPLE", False), Gdk.Atom.intern("SAVE_TARGETS", False), Gdk.Atom.intern("UTF8_STRING", False), Gdk.Atom.intern("COMPOUND_TEXT", False), Gdk.Atom.intern("TEXT", False), Gdk.Atom.intern("STRING", False), Gdk.Atom.intern("text/plain;charset=utf-8", False), Gdk.Atom.intern("text/plain", False)], we want parsetree``

case ==>  ``file://etc``

* Copy/cut, then paste does NOT cause paste into same page or different page. No action or message occurs with either <ctrl>c or context menu [<right click>paste ]. Context menu paste is not greyed out.
* Copied text from zim exists in system clipboard (clipboard paste into external editor) w/o zim formatting (i.e., no html link) even if copied text is formatted on zim page as a link and copied as link.
* [``/tmp/zim-<random``](file:///tmp/zim-%3Crandom)``>/zim.log relevant text after copy/paste:``

``DEBUG: Targets available for paste: [Gdk.Atom.intern("TIMESTAMP", False), Gdk.Atom.intern("TARGETS", False), Gdk.Atom.intern("MULTIPLE", False), Gdk.Atom.intern("SAVE_TARGETS", False), Gdk.Atom.intern("UTF8_STRING", False), Gdk.Atom.intern("COMPOUND_TEXT", False), Gdk.Atom.intern("TEXT", False), Gdk.Atom.intern("STRING", False), Gdk.Atom.intern("text/plain;charset=utf-8", False), Gdk.Atom.intern("text/plain", False)], we want parsetree``
``DEBUG: Requesting data for ('text/x-zim-parsetree', 0, 1) -- using set_with_data() workaround``
``	Traceback (most recent call last):``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/gui/pageview/__init__.py", line 4137, in do_paste_clipboard``
``	    self.get_buffer().paste_clipboard(Clipboard, None, self.get_editable(), text_format=format)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/gui/pageview/__init__.py", line 3372, in paste_clipboard``
``	    parsetree = clipboard.get_parsetree(self.notebook, self.page, text_format)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/gui/clipboard.py", line 641, in get_parsetree``
``	    return parsetree_from_selectiondata(selectiondata, notebook, path, text_format)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/gui/clipboard.py", line 227, in parsetree_from_selectiondata``
``	    parsetree = ParseTree().fromstring(selectiondata.get_data())``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/gui/clipboard.py", line 377, in get_data``
``	    return self.data.get_data_as(target_id)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/gui/clipboard.py", line 460, in get_data_as``
``	    set_parsetree_attributes_to_resolve_links(self.parsetree, self.notebook, self.path)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/notebook/content_updater.py", line 34, in set_parsetree_attributes_to_resolve_links``
``	    parsetree.replace(``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/formats/__init__.py", line 668, in replace``
``	    self._replace(self._etree.getroot(), tags, func)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/formats/__init__.py", line 685, in _replace``
``	    self._replace(child, tags, func) # recurs``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/formats/__init__.py", line 679, in _replace``
``	    replacement = func(Element.new_from_etree(child))``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/notebook/content_updater.py", line 48, in _resolve_links_and_images``
``	    target = notebook.resolve_file(href, src_path)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/notebook/notebook.py", line 1054, in resolve_file``
``	    return File(filename)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/fs.py", line 1052, in __init__``
``	    FilePath.__init__(self, path)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/fs.py", line 366, in __init__``
``	    path = self._parse_uri(path)``
``	  File "/home/user1001/.local/lib/zim-0.74.0/zim/fs.py", line 395, in _parse_uri``
``	    assert False, 'Can not handle non-local file uris``'

* Context menu via right click on text ``"file://etc" ``(when formatted as HTML link) shows options appropriate for 'no link' text. There are no options to add, edit, or remove link. [possible issue]
* mouse pointer changes to link pointer when hovered over file URI (and pop-up is displayed) even though context menu does not have link menu items (copy link, remove link, edit link) [possible issue]
* 'edit source' shows wiki formatting around file URI even though context menu would indicate no link formatted. 
* clear formatting (<ctrl>9) on file URI clears HTML link formatting.
* pop up shows ``"open smb://etc" [possible issue]``
* mouse click on file URI produces zim bug message:

	This is zim 0.74.0
	Platform: posix
	Locale: en_US UTF-8
	FS encoding: utf-8
	Python: (3, 9, 5, 'final', 0)
	PyGObject: (3, 38, 0)

	
``======= Traceback =======``
``  File "/home/user1001/.local/lib/zim-0.74.0/zim/gui/pageview/__init__.py", line 6501, in do_activate_link``
``    self._do_activate_link(link, hints)``
``  File "/home/user1001/.local/lib/zim-0.74.0/zim/gui/pageview/__init__.py", line 6514, in _do_activate_link``
``    path = self.notebook.resolve_file(link, self.page)``
``  File "/home/user1001/.local/lib/zim-0.74.0/zim/notebook/notebook.py", line 1054, in resolve_file``
``    return File(filename)``
``  File "/home/user1001/.local/lib/zim-0.74.0/zim/fs.py", line 1052, in __init__``
``    FilePath.__init__(self, path)``
``  File "/home/user1001/.local/lib/zim-0.74.0/zim/fs.py", line 366, in __init__``
``    path = self._parse_uri(path)``
``  File "/home/user1001/.local/lib/zim-0.74.0/zim/fs.py", line 395, in _parse_uri``
``    assert False, 'Can not handle non-local file uris``'
``AssertionError: Can not handle non-local file uris``

case ==>  ``file:///etc``

* is the original text obtained via dragging system folder ``/etc`` onto a zim page. zim created the file URI link from the drag-and-drop operation
* after copy/cut, then attempted paste does NOT paste into different page. No action or message occurs with either <ctrl>c or context menu [<right click>paste ]. Context menu paste is not greyed out.
* Copied text from zim exists in system clipboard w/o zim formatting (no html link) even if copied text is formatted on zim page as a link. e.g., problematic text can be copied from 'edit source' from the source zim page then pasted via 'edit source' on target zim page. 
* DOES paste into same page as "``file:///etc"`` with <ctrl>c or context menu [<right click>paste ] into another zim page with auto-link added when copied as link or not link or as verbatim format
* pop up shows ``"open file:///etc" ``
* mouse click on file URI opens file browser (e.g., thunar) at location ``/etc``
* [``/tmp/zim-<random``](file:///tmp/zim-%3Crandom)``>/zim.log relevant text after copy/paste:``

``DEBUG: Targets available for paste: [Gdk.Atom.intern("TIMESTAMP", False), Gdk.Atom.intern("TARGETS", False), Gdk.Atom.intern("MULTIPLE", False), Gdk.Atom.intern("SAVE_TARGETS", False), Gdk.Atom.intern("UTF8_STRING", False), Gdk.Atom.intern("COMPOUND_TEXT", False), Gdk.Atom.intern("TEXT", False), Gdk.Atom.intern("STRING", False), Gdk.Atom.intern("text/plain;charset=utf-8", False), Gdk.Atom.intern("text/plain", False)], we want parsetree``
``DEBUG: Requesting data for ('text/x-zim-parsetree', 0, 1) -- using set_with_data() workaround``

case ==>  ``smb://etc``

* included as case because of ``smb://`` popup display for file URI ``file://``
* copy/paste works as expected
* mouse click on URI opens default browser
* pop up shows ``"open smb://etc"``

case==>  ``/etc``

* copy/paste works as expected
* mouse click on URI opens file browser to ``/etc``

**Comment by jaap-karssenberg** [comment]
Will do some testing to see if I can reproduce the issue.

One clarifying question: when you talk about "file:etc" URI, did you mean
"file:/" or "file://" ? Without the "/" Zim will recognize the link as a
page called "file" not an URI. You show an error with non-local files,
which makes me suspect you are using a version of the "file:/" URI scheme
were a host name is expected.

Regards,

Jaap


Op zo 19 sep. 2021 22:56 schreef ZtzEAcRsND ***@***.***>:

**Comment by ZtzEAcRsND** [comment]
In the "More Detail" discussion, I meant both "file:" and "file:/" even though the issues are with "file://" and "file:///".  Initially, I found the issue trying to copy a paragraph with a local link created drag and drop of "/etc" which converted to "file:///".  The core issue is inability to paste "file://" and "file:///". 

After figuring out what caused the failure to paste by incrementally removing parts of the paragraph, I manually tried cut/copy/paste all variations (file: , file:/ , file://, file:/// ) even though (e.g.,)  "file:" should be wiki syntax and "file:///" is a non-local file URI.  In the original issue text,  these tests on the various forms are indicated by the "case ==>". In the process of testing, I noticed other "possible issues" not related to cut/copy/paste, e.g.,  a` link formatted "file:/etc" click links to "file:etc" per zim wiki syntax, but the pop up display shows "file:///etc", supposedly a non-local URI and  (e.g.,) mouse-click on a link-formatted "file://etc" causes a "found a bug" pop up.

Let me know if I should break out and submit separately the other "possible issues". 
...and thanks.
